lib/SILOptimizer/Mandatory/TFConstExpr.cpp

```diff
@@ -824,29 +930,56 @@ ConstExprFunctionCache::evaluateFlowSensitive(SILInstruction *inst) {
   if (auto apply = dyn_cast<ApplyInst>(inst))
     return computeCallResult(apply);
 
-  if (auto *store = dyn_cast<StoreInst>(inst)) {
-    auto stored = getConstantValue(inst->getOperand(0));
-    if (!stored.isConstant())
-      return stored;
 
+  // We have a couple forms of store instruction, this is the logic for handling
+  // storing of a constant to memory.
+  auto evaluateFSStore = [&](SymbolicValue storedCst, SILValue dest) ->
+                               llvm::Optional<SymbolicValue> {
     // Only update existing memory locations that we're tracking.
-    auto it = calculatedValues.find(inst->getOperand(1));
+    auto it = calculatedValues.find(dest);
     if (it == calculatedValues.end())
       return SymbolicValue::getUnknown(inst, UnknownReason::Default);
 
-    // If this is a store to an address, update the element of the base value.
-    if (it->second.isAddress()) {
-      auto baseVal = getConstantValue(it->second.getAddressBase());
-      auto indices = it->second.getAddressIndices();
-
-      if (updateIndexedElement(baseVal, indices, stored,
-                               evaluator.getAllocator()))
-        return SymbolicValue::getUnknown(inst, UnknownReason::Default);
-      stored = baseVal;
+    // If this is a direct store to tracked memory object, just update it.
+    if (!it->second.isAddress()) {
+      it = calculatedValues.find(dest);
+      assert(it != calculatedValues.end());
+      it->second = storedCst;
+      return None;
     }
 
-    it->second = stored;
+    // Otherwise, this is a store to a derived address, update the element of
+    // the base value.
+    auto base = it->second.getAddressBase();
+    auto baseVal = getConstantValue(base);
+    auto baseType = simplifyType(base->getType().getSwiftRValueType());
+    auto indices = it->second.getAddressIndices();
+
+    if (updateIndexedElement(baseVal, indices, storedCst, baseType,
+                             evaluator.getAllocator()))
+      return SymbolicValue::getUnknown(inst, UnknownReason::Default);
+
+    it = calculatedValues.find(base);
+    assert(it != calculatedValues.end());
+    it->second = baseVal;
     return None;
+  };
+
+  if (auto *store = dyn_cast<StoreInst>(inst)) {
+    auto stored = getConstantValue(inst->getOperand(0));
+    if (!stored.isConstant())
+      return stored;
+
+    return evaluateFSStore(stored, inst->getOperand(1));
+  }
+
+  // Copy addr is a load + store combination.
+  if (auto *copy = dyn_cast<CopyAddrInst>(inst)) {
```

Comment:
ACAICT, the number of `dyn_cast` cases is growing within `ConstExprFunctionCache::evaluateFlowSensitive`. Although a visitor is too heavyweight for this, is it worth turning them into a switch stmt? You didn't use methods specific to subclasses, so a switch seems to work better.

Reply:
This is the standard pattern.  Are you concerned about compile time overheads due to repeated if/else chains?   Years ago I did work to make sure the LLVM optimizer turned this into a switch on opcode.  It is possible that someone broke it though.

Reply:
K! That's good to know.